core/iwasm/interpreter/wasm_opcode.h

```diff
@@ -686,230 +686,230 @@ typedef enum WASMAtomicEXTOpcode {
-        HANDLE_OPCODE(WASM_OP_F32_CONVERT_S_I64),    /* 0xb4 */ \
-        HANDLE_OPCODE(WASM_OP_F32_CONVERT_U_I64),    /* 0xb5 */ \
-        HANDLE_OPCODE(WASM_OP_F32_DEMOTE_F64),       /* 0xb6 */ \
-        HANDLE_OPCODE(WASM_OP_F64_CONVERT_S_I32),    /* 0xb7 */ \
-        HANDLE_OPCODE(WASM_OP_F64_CONVERT_U_I32),    /* 0xb8 */ \
-        HANDLE_OPCODE(WASM_OP_F64_CONVERT_S_I64),    /* 0xb9 */ \
-        HANDLE_OPCODE(WASM_OP_F64_CONVERT_U_I64),    /* 0xba */ \
-        HANDLE_OPCODE(WASM_OP_F64_PROMOTE_F32),      /* 0xbb */ \
-        HANDLE_OPCODE(WASM_OP_I32_REINTERPRET_F32),  /* 0xbc */ \
-        HANDLE_OPCODE(WASM_OP_I64_REINTERPRET_F64),  /* 0xbd */ \
-        HANDLE_OPCODE(WASM_OP_F32_REINTERPRET_I32),  /* 0xbe */ \
-        HANDLE_OPCODE(WASM_OP_F64_REINTERPRET_I64),  /* 0xbf */ \
-        HANDLE_OPCODE(WASM_OP_I32_EXTEND8_S),        /* 0xc0 */ \
-        HANDLE_OPCODE(WASM_OP_I32_EXTEND16_S),       /* 0xc1 */ \
-        HANDLE_OPCODE(WASM_OP_I64_EXTEND8_S),        /* 0xc2 */ \
-        HANDLE_OPCODE(WASM_OP_I64_EXTEND16_S),       /* 0xc3 */ \
-        HANDLE_OPCODE(WASM_OP_I64_EXTEND32_S),       /* 0xc4 */ \
-        HANDLE_OPCODE(WASM_OP_DROP_64),              /* 0xc5 */ \
-        HANDLE_OPCODE(WASM_OP_SELECT_64),            /* 0xc6 */ \
-        HANDLE_OPCODE(EXT_OP_GET_LOCAL_FAST),        /* 0xc7 */ \
-        HANDLE_OPCODE(EXT_OP_SET_LOCAL_FAST_I64),    /* 0xc8 */ \
-        HANDLE_OPCODE(EXT_OP_SET_LOCAL_FAST),        /* 0xc9 */ \
-        HANDLE_OPCODE(EXT_OP_TEE_LOCAL_FAST),        /* 0xca */ \
-        HANDLE_OPCODE(EXT_OP_TEE_LOCAL_FAST_I64),    /* 0xcb */ \
-        HANDLE_OPCODE(EXT_OP_COPY_STACK_TOP),        /* 0xcc */ \
-        HANDLE_OPCODE(EXT_OP_COPY_STACK_TOP_I64),    /* 0xcd */ \
-        HANDLE_OPCODE(EXT_OP_COPY_STACK_VALUES),     /* 0xce */ \
-        HANDLE_OPCODE(WASM_OP_IMPDEP),               /* 0xcf */ \
-        HANDLE_OPCODE(WASM_OP_REF_NULL),             /* 0xd0 */ \
-        HANDLE_OPCODE(WASM_OP_REF_IS_NULL),          /* 0xd1 */ \
-        HANDLE_OPCODE(WASM_OP_REF_FUNC),             /* 0xd2 */ \
-        HANDLE_OPCODE(EXT_OP_BLOCK),                 /* 0xd3 */ \
-        HANDLE_OPCODE(EXT_OP_LOOP),                  /* 0xd4 */ \
-        HANDLE_OPCODE(EXT_OP_IF),                    /* 0xd5 */ \
-        HANDLE_OPCODE(EXT_OP_BR_TABLE_CACHE),        /* 0xd6 */ \
-    };                                                          \
-    do {                                                        \
-        _name[WASM_OP_MISC_PREFIX] =                            \
-            HANDLE_OPCODE(WASM_OP_MISC_PREFIX); /* 0xfc */      \
-        _name[WASM_OP_ATOMIC_PREFIX] =                          \
-            HANDLE_OPCODE(WASM_OP_ATOMIC_PREFIX); /* 0xfe */    \
-        DEF_DEBUG_BREAK_HANDLE(_name)                           \
+#define DEFINE_GOTO_TABLE(type, _name)                                    \
+    static os_thread_local_attribute type _name[WASM_INSTRUCTION_NUM] = { \
```

Comment:
is there a data race risk here? I thought it's only initialized once before first access, and it's read-only.

Reply:
It's not done only once right now https://github.com/bytecodealliance/wasm-micro-runtime/blob/739acfc9082e63967ee91fca85dc01ab1630a5cc/core/iwasm/interpreter/wasm_interp_classic.c#L1170
but I'll see if I can move the initialization somewhere else (also because I noticed from the CI that TLS is not supported on all platforms)

Reply:
It is a `static const void * handle_table[256]` table, its data isn't changed after it is defined, is there any data race risk on it?

Reply:
```bash
WARNING: ThreadSanitizer: data race (pid=1397466)
  Read of size 8 at 0x55f7ae9d2570 by main thread:
    #0 wasm_interp_call_func_bytecode wasm-micro-runtime/core/iwasm/interpreter/wasm_interp_classic.c:2123 (iwasm+0x4f629)
    #1 wasm_interp_call_wasm wasm-micro-runtime/core/iwasm/interpreter/wasm_interp_classic.c:4207 (iwasm+0x5a062)
    #2 call_wasm_with_hw_bound_check wasm-micro-runtime/core/iwasm/interpreter/wasm_runtime.c:2261 (iwasm+0x1c01e)
[...]

  Previous write of size 8 at 0x55f7ae9d2570 by thread T1:
    #0 wasm_interp_call_func_bytecode wasm-micro-runtime/core/iwasm/interpreter/wasm_interp_classic.c:1170 (iwasm+0x49feb)
    #1 wasm_interp_call_wasm wasm-micro-runtime/core/iwasm/interpreter/wasm_interp_classic.c:4207 (iwasm+0x5a062)
    #2 call_wasm_with_hw_bound_check wasm-micro-runtime/core/iwasm/interpreter/wasm_runtime.c:2261 (iwasm+0x1c01e)
[...]
```
I guess the problem is with what is done after the static initialization, in the `do ... while` https://github.com/bytecodealliance/wasm-micro-runtime/blob/739acfc9082e63967ee91fca85dc01ab1630a5cc/core/iwasm/interpreter/wasm_opcode.h#L908

Reply:
I moved everything into the initialization, that seems to fix the problem